include/boost/multiprecision/cpp_int/cpp_int_config.hpp

```diff
@@ -17,57 +20,80 @@ namespace detail {
-// [unsigned] boost::long_long_type, long, int, which has the specified number
-// of bits.  Note that intN_t and boost::int_t<N> find the first
+// [unsigned] long long, long, int, which has the specified number
+// of bits.  Note that int_t and uint_t find the first
 // member of the above list, not the last.  We want the last in the
 // list to ensure that mixed arithmetic operations are as efficient
 // as possible.
 //
+
+template <unsigned Bits>
+struct int_t
+{
+   using exact = typename std::conditional<Bits == sizeof(signed char) * CHAR_BIT, signed char,
+                 typename std::conditional<Bits == sizeof(short) * CHAR_BIT, short,
+                 typename std::conditional<Bits == sizeof(int) * CHAR_BIT, int,
+                 typename std::conditional<Bits == sizeof(long) * CHAR_BIT, long, long long
+                 >::type>::type>::type>::type;
+
+};
+
+template <unsigned Bits>
+struct uint_t
+{
+   using exact = typename std::conditional<Bits == sizeof(unsigned char) * CHAR_BIT, unsigned char,
+                 typename std::conditional<Bits == sizeof(unsigned short) * CHAR_BIT, unsigned short,
+                 typename std::conditional<Bits == sizeof(unsigned int) * CHAR_BIT, unsigned int,
+                 typename std::conditional<Bits == sizeof(unsigned long) * CHAR_BIT, unsigned long, unsigned long long
+                 >::type>::type>::type>::type;
+
+};
+
 template <unsigned N>
 struct largest_signed_type
 {
    using type = typename std::conditional<
-       1 + std::numeric_limits<boost::long_long_type>::digits == N,
-       boost::long_long_type,
+       1 + std::numeric_limits<long long>::digits == N,
+       long long,
        typename std::conditional<
            1 + std::numeric_limits<long>::digits == N,
            long,
            typename std::conditional<
                1 + std::numeric_limits<int>::digits == N,
                int,
-               typename boost::int_t<N>::exact>::type>::type>::type;
+               typename int_t<N>::exact>::type>::type>::type;
 };
 
 template <unsigned N>
 struct largest_unsigned_type
 {
    using type = typename std::conditional<
-       std::numeric_limits<boost::ulong_long_type>::digits == N,
-       boost::ulong_long_type,
+       std::numeric_limits<unsigned long long>::digits == N,
+       unsigned long long,
        typename std::conditional<
            std::numeric_limits<unsigned long>::digits == N,
            unsigned long,
            typename std::conditional<
                std::numeric_limits<unsigned int>::digits == N,
                unsigned int,
-               typename boost::uint_t<N>::exact>::type>::type>::type;
+               typename uint_t<N>::exact>::type>::type>::type;
 };
 
 } // namespace detail
 
 #if defined(BOOST_HAS_INT128)
 
 using limb_type = detail::largest_unsigned_type<64>::type;
-using signed_limb_type = detail::largest_signed_type<64>::type  ;
-using double_limb_type = boost::uint128_type                    ;
```

Comment:
Using these typedefs suppresses endless warnings from the compiler about using a non-std extension - perhaps we can lift that logic from Boost.Config?

Reply:
That should be easy enough. I think I need to define `BOOST_HAS_INT128` in a standalone config like math has.